### DISCUSSION TITLE
[yargs] Access to command result is broken

### DISCUSSION THREAD

**Discussion by orgads** [DefinitelyTyped/DefinitelyTyped#63990] [Issues with a @types package]
Regression between 17.0.19 and 17.0.20.

Broke by:
- #63694

```typescript
import * as yargs from 'yargs';

const argv = yargs
  .command('foo', 'bar', (args) => args
    .option('test', { coerce: (test) => test as string[] })
  ).parseSync();

console.log(argv.test.join(',')); // Property 'join' does not exist on type 'unknown'
```

On another project the error is `Property 'join' does not exist on type '{}'`.

[a comment by an automated account is not shown]

**Reply by gerben-stavenga** [reply]
In a sense it's now WAI. The problem here is that argv.test would ONLY exist if the command foo was given otherwise. So giving it the type string[] would violate the typescript rules. The type here is dynamic depending on the input data and not static.

**Reply by orgads** [reply]
This is a simplified example. Our real use-case is accepting either an array or a string, which it splits:

```typescript
coerce: (sections) => (Array.isArray(sections) ? sections : sections?.split(',')) as string[]
```

**Reply by gerben-stavenga** [reply]
It's not about coerce, it's about how yargs evaluates the command line. The option "test" is only parsed if the command is foo.

Ie.

```
program foo -test bla    => argv.test = "bla"
program baz -test bla   => argv.test = undefined
```

The option "test" is only parsed as an option in the context of command foo. That's why the argv variable cannot be statically typed to have field test.

If you want the option "test" unconditionally evaluated then you can do

```
const argv = yargs.option('test', { coerce: (test) => test as string[] }).parseSync();

console.log(argv.test.join(','));
```

**Reply by orgads** [reply]
Thank you.

I have a single command, and I can't use the top-level yargs, because I'm using `positional`.

```typescript
const argv = yargs
  .command('$0 <filename>', 'Loads configuration from file.', (args) => args
    .positional('filename', {
      describe: 'Configuration file',
      type: 'string'
    })
  // ...
```

Can you suggest how to resolve this?

**Reply by gerben-stavenga** [reply]
ideally you write an arrow function invoked by command
```
const argv = yargs
  .command(
      '$0 <filename>', 'Loads configuration from file.', 
      (yargs) => yargs
            .option('test', { coerce: (test) => test as string[] })
            .positional('filename', {
              describe: 'Configuration file',
              type: 'string'
            }),
     async (argv) => { /* your code here where argv has option test */ }
  )
```

Because yargs is invoking the arrow function when the it actually evaluates command we know for sure that test option is present.

**Comment by pmconne** [comment]
We're experiencing a similar regression when upgrading from 17.0.19 to 17.0.20.
Every use of the `command` function overload taking a `CommandModule` in [this file](https://github.com/iTwin/itwinjs-core/blob/e4936aa58a3c13fc25863dda6d467a51defad4b3/utils/workspace-editor/src/WorkspaceEditor.ts#L604) now complains that we need to supply 2-6 arguments instead of the 1 that was previously accepted.
```
src/WorkspaceEditor.ts(604,7): error TS2554: Expected 2-6 arguments, but got 1.
/* ...18 more occurrences... */
src/WorkspaceEditor.ts(714,7): error TS2554: Expected 2-6 arguments, but got 1.
```

**Reply by ZeeshanTamboli** [reply]
@pmconne I think you have to remove the `AddFileOptions` generics. I had to do the same in Material UI.

**Reply by Shingyx** [reply]
FYI I'm looking at fixing this `CommandModule` regression in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66725 (now merged)

**Comment by sybereal** [comment]
I've experienced another regression regarding `CommandModule`. It tells me
```
Types of property 'handler' are incompatible. [...]
Type 'ArgumentsCamelCase<any>' is not assignable to type 'ArgumentsCamelCase<ExpectedArgs>'.
Type 'ArgumentsCamelCase<any>' is missing the following properties from type {...}: [list of options added by module builder function]
```

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4d9825d39eac3b34de18240762378151debb3941/types/yargs/index.d.ts#L178
If I replace the `CommandModule<T, any>` in this line with `CommandModule<T, U>`  and re-add the generic parameter, it works again.
N.B.: I am not suggesting to change the return type back to `Argv<U>`. I understand why that was wrong.

**Reply by Shingyx** [reply]
FYI I'm looking at fixing this `CommandModule` regression in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66725 (now merged)
